py/objslice.c

```diff
@@ -53,29 +55,114 @@ STATIC void slice_print(const mp_print_t *print, mp_obj_t o_in, mp_print_kind_t
     mp_print_str(print, ")");
 }
 
+#if MICROPY_PY_BUILTINS_SLICE_INDICES
+STATIC mp_obj_t slice_indices(mp_obj_t self_in, mp_obj_t length_obj) {
+    mp_obj_slice_t *self = MP_OBJ_TO_PTR(self_in);
+    mp_int_t length = mp_obj_int_get_checked(length_obj);
+    mp_int_t start, stop, step;
+    mp_obj_t results[3]; // In start, stop, step order
+
+    if (self->step == mp_const_none) {
+        step = 1;
+    } else {
+        step = mp_obj_int_get_checked(self->step);
+        if (step == 0) {
+            mp_raise_ValueError("slice step cannot be zero");
+        }
+    }
+
+    if (step > 0) {
+        // Positive step
+        if (self->start == mp_const_none) {
+            start = 0;
+        } else {
+            start = mp_obj_int_get_checked(self->start);
+            if (start < 0) {
+                start += length;
+            }
+            start = MIN(length, MAX(start, 0));
+        }
+
+        if (self->stop == mp_const_none) {
+            stop = length;
+        } else {
+            stop = mp_obj_int_get_checked(self->stop);
+            if (stop < 0) {
+                stop += length;
+            }
+            stop = MIN(length, MAX(stop, 0));
+        }
+    } else {
+        // Negative step
+        if (self->start == mp_const_none) {
+            start = length - 1;
+        } else {
+            start = mp_obj_int_get_checked(self->start);
+            if (start < 0) {
+                start += length;
+            }
+            start = MIN(length-1, MAX(start, -1));
+        }
+
+        if (self->stop == mp_const_none) {
+            stop = -1;
+        } else {
+            stop = mp_obj_int_get_checked(self->stop);
+            if (stop < 0) {
+                stop += length;
+            }
+            stop = MIN(length-1, MAX(stop, -1));
+        }
+    }
+
+    results[0] = MP_OBJ_NEW_SMALL_INT(start);
+    results[1] = MP_OBJ_NEW_SMALL_INT(stop);
+    results[2] = MP_OBJ_NEW_SMALL_INT(step);
+
+    return mp_obj_new_tuple(3, results);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_2(slice_indices_obj, slice_indices);
+#endif
+
 #if MICROPY_PY_BUILTINS_SLICE_ATTRS
 STATIC void slice_attr(mp_obj_t self_in, qstr attr, mp_obj_t *dest) {
     if (dest[0] != MP_OBJ_NULL) {
         // not load attribute
         return;
     }
     mp_obj_slice_t *self = MP_OBJ_TO_PTR(self_in);
+
     if (attr == MP_QSTR_start) {
         dest[0] = self->start;
     } else if (attr == MP_QSTR_stop) {
         dest[0] = self->stop;
     } else if (attr == MP_QSTR_step) {
         dest[0] = self->step;
+#if MICROPY_PY_BUILTINS_SLICE_INDICES
+    } else if (attr == MP_QSTR_indices) {
+        dest[0] = mp_obj_new_bound_meth(MP_OBJ_FROM_PTR(&slice_indices_obj),
+                                        self_in);
```

Comment:
To avoid allocation of a bound-method object this can be optimised to:
```c
dest[0] = MP_OBJ_FROM_PTR(&slice_indices_obj);
dest[1] = self_in;
```

Reply:
OK, thanks for that. I wasn't quite sure how to do that from the `slice_attr` function.